engine/context/src/main/java/io/deephaven/engine/context/QueryCompiler.java

```diff
@@ -53,24 +53,85 @@ public class CompilerTools {
 
     private static final String IDENTIFYING_FIELD_NAME = "_CLASS_BODY_";
 
-    private static final String CODEGEN_TIMEOUT_PROP = "CompilerTools.codegen.timeoutMs";
+    private static final String CODEGEN_TIMEOUT_PROP = "QueryCompiler.codegen.timeoutMs";
     private static final long CODEGEN_TIMEOUT_MS_DEFAULT = TimeUnit.SECONDS.toMillis(10); // 10 seconds
-    private static final String CODEGEN_LOOP_DELAY_PROP = "CompilerTools.codegen.retry.delay";
+    private static final String CODEGEN_LOOP_DELAY_PROP = "QueryCompiler.codegen.retry.delay";
     private static final long CODEGEN_LOOP_DELAY_MS_DEFAULT = 100;
     private static final long codegenTimeoutMs =
             Configuration.getInstance().getLongWithDefault(CODEGEN_TIMEOUT_PROP, CODEGEN_TIMEOUT_MS_DEFAULT);
     private static final long codegenLoopDelayMs =
             Configuration.getInstance().getLongWithDefault(CODEGEN_LOOP_DELAY_PROP, CODEGEN_LOOP_DELAY_MS_DEFAULT);
 
+    private static boolean logEnabled = Configuration.getInstance().getBoolean("QueryCompiler.logEnabledDefault");
+
+    public static final String FORMULA_PREFIX = "io.deephaven.temp";
+    public static final String DYNAMIC_GROOVY_CLASS_PREFIX = "io.deephaven.dynamic";
+
+    public static QueryCompiler create(File cacheDirectory, ClassLoader classLoader) {
+        return new QueryCompiler(cacheDirectory, classLoader, true);
+    }
+
+    static QueryCompiler createForUnitTests() {
+        return new QueryCompiler(new File(Configuration.getInstance().getWorkspacePath() +
+                File.separator + "cache" + File.separator + "classes"));
+    }
+
+    private final Map<String, CompletableFuture<Class<?>>> knownClasses = new HashMap<>();
+
+    private final String[] dynamicPatterns = new String[] {DYNAMIC_GROOVY_CLASS_PREFIX, FORMULA_PREFIX};
+
+    private final File classDestination;
+    private final boolean isCacheDirectory;
+    private final Set<File> additionalClassLocations;
+    private volatile WritableURLClassLoader ucl;
+
+    /** package-private constructor for {@link io.deephaven.engine.context.PoisonedQueryCompiler} */
+    QueryCompiler() {
+        classDestination = null;
+        isCacheDirectory = false;
+        additionalClassLocations = null;
+    }
+
+    private QueryCompiler(File classDestination) {
+        this(classDestination, null, false);
+    }
+
+    private QueryCompiler(File classDestination, ClassLoader parentClassLoader, boolean isCacheDirectory) {
+        if (parentClassLoader == null) {
+            parentClassLoader = this.getClassLoader();
```

Comment:
I think this is incorrect. `this.getClassLoader()` returns an as-yet-uninitialized instance member, `ucl`. Maybe you want `QueryCompiler.class.getClassLoader()`?

Also, I slightly prefer the following pattern:
1. Declare `parentClassLoader` `final`.
2. Create `final ClassLoader parentClassLoaderToUse = parentClassLoader != null ? parentClassLoader : QueryCompiler.class.getClassLoader();`